vscode-lean4/src/utils/clientProvider.ts

```diff
@@ -103,29 +116,68 @@ export class LeanClientProvider implements Disposable {
     }
 
     async didOpenEditor(document: TextDocument) {
-        // All open .lean files are assumed to be Lean 4 files.
-        // We need to do this because by default, .lean is associated with language id `lean`,
-        // i.e. Lean 3. vscode-lean is expected to yield when isLean4Project is true.
-        if (document.languageId === 'lean') {
-            // Only change the document language for *visible* documents,
-            // because this closes and then reopens the document.
-            await languages.setTextDocumentLanguage(document, 'lean4');
-        } else if (document.languageId !== 'lean4') {
+        // bail as quickly as possible on non-lean files.
+        if (document.languageId !== 'lean' && document.languageId !== 'lean4') {
             return;
         }
 
-        const client = await this.ensureClient(document.uri, null);
-        await client.openLean4Document(document)
+        if (!this.getVisibleEditor(document.uri)) {
+            // Sometimes VS code opens a document that has no editor yet.
+            // For example, this happens when the vs code opens files to get git
+            // information using a "git:" Uri scheme:
+            //  git:/d%3A/Temp/lean_examples/Foo/Foo/Hello.lean.git?%7B%22path%22%3A%22d%3A%5C%5CTemp%5C%5Clean_examples%5C%5CFoo%5C%5CFoo%5C%5CHello.lean%22%2C%22ref%22%3A%22%22%7D
+            return;
+        }
+
+        try {
+            // All open .lean files are assumed to be Lean 4 files.
+            // We need to do this because by default, .lean is associated with language id `lean`,
+            // i.e. Lean 3. vscode-lean is expected to yield when isLean4Project is true.
+            if (document.languageId === 'lean') {
+                // Only change the document language for *visible* documents,
+                // because this closes and then reopens the document.
+                await languages.setTextDocumentLanguage(document, 'lean4');
+
+                // setTextDocumentLanguage triggers another didOpenEditor event,
+                // and we want that event callback to be the one that calls
+                // ensureClient, so we bail here so we don't try and do it twice
+                // on the same document.
+                return;
```

Comment:
I vaguely remember that I didn't get the second didOpenEditor event when I tried it last time.  (That's why the condition was fall-through.)

Reply:
seems reliable now, but we can keep an eye on if, if you find a repro for when the second open event does not happen let me know...